src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java

```diff
@@ -189,13 +189,16 @@ public static class Execution extends AbstractStepExecutionImpl {
             Object returnValue = null;
             Throwable error = null;
             boolean success = false, failure = false, sync = true;
+            String c = Jenkins.XSTREAM.toXML(getContext());
             synchronized (s) {
                 if (s.returnValues.containsKey(k)) {
                     success = true;
                     returnValue = s.returnValues.get(k);
                 } else if (s.errors.containsKey(k)) {
                     failure = true;
                     error = s.errors.get(k);
+                } else {
+                    s.contexts.put(k, c);
```

Comment:
This is the fix. Other threads now only ever see `State` in one of two modes: either `contexts` does not contain the relevant key, and `returnValues` and `errors` should be used and will be checked here, or `contexts` does contain the key, and the `StepContext` should be used instead. Previously, there was a case where `contexts` did not contain the key but `returnValues` and `errors` had already been checked.